python/smqtk/iqr/iqr_session.py

```diff
@@ -324,8 +361,109 @@ def reset(self):
+        z.writestr(self.STATE_ZIP_FILENAME, json.dumps({
+            'pos': pos_d,
+            'neg': neg_d,
+            'external_pos': [(self._query_uuid,ext_pos_d[self._query_uuid].type(),self._query_f.tolist())],
+            'external_neg': ext_neg_d,
+            'query_f':self._query_f.tolist(),
+            'query_uuid':self._query_uuid
+        }))
+        z.close()
+        return z_buffer.getvalue()
+
+    def set_state_bytes(self, b, descriptor_factory):
+        """
+        Set this session's state to the given byte representation, resetting
+        this session in the process.
+        Bytes given must have been retrieved via a previous call to
+        ``get_state_bytes`` otherwise this method will fail.
+        Since this state may be completely different from the current state,
+        this session is reset before applying the new state. Thus, any current
+        ranking results are thrown away.
+        :param b: Bytes to set this session's state to.
+        :type b: bytes
+        :param descriptor_factory: Descriptor element factory to use when
+            generating descriptor elements from extracted data.
+        :type descriptor_factory: smqtk.representation.DescriptorElementFactory
+        :raises ValueError: The input bytes could not be loaded due to
+            incompatibility.
+        """
+        z_buffer = io.BytesIO(b)
+        z = zipfile.ZipFile(z_buffer, 'r', self.STATE_ZIP_COMPRESSION)
+        if self.STATE_ZIP_FILENAME not in z.namelist():
+            raise ValueError("Invalid bytes given, did not contain expected "
+                             "zipped file name.")
+
+        # Extract expected json file object
+        state = json.loads(z.read(self.STATE_ZIP_FILENAME).decode())
+        del z, z_buffer
+
+        with self:
+            self.reset()
+
+            def load_descriptor(_uid, _type_str, vec_list):
+                _e = descriptor_factory.new_descriptor(_type_str, _uid)
+                if _e.has_vector():
+                    assert _e.vector().tolist() == vec_list, \
+                        "Found existing vector for UUID '%s' but vectors did " \
+                        "not match."
+                else:
+                    _e.set_vector(vec_list)
+                return _e
+
+            # Read in raw descriptor data from the state, convert to descriptor
+            # element, then store in our descriptor sets.
+            for source, target in [(state['external_pos'],
+                                    self.external_positive_descriptors),
+                                   (state['external_neg'],
+                                    self.external_negative_descriptors),
+                                   (state['pos'], self.positive_descriptors),
+                                   (state['neg'], self.negative_descriptors)]:
+                for uid, type_str, vector_list in source:
+                    e = load_descriptor(uid, type_str, vector_list)
```

Comment:
A more error-robust approach to this loop would be to load all descriptors, collecting them in a local list, before adding them to / updating the ``target`` set. That way if any descriptor load produces an error (the assert in the ``load_descriptor`` function) then the ``target`` set is not modified. Currently, the ``target`` set can end up in a partially modified state if there is an error part-way through.

Reply:
Removed load_descriptor and squashed its operations into the for loop so the assertion error can be caught by a try block